src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/HandlerMethodDetails.cs

```diff
@@ -0,0 +1,185 @@
+            const string HandleNotificationSuffix = "." + HandleNotificationName;
+
+            foreach (var methodInfo in handlerType.GetRuntimeMethods())
+            {
+                if (MethodInfoMatches(methodInfo))
+                    return methodInfo.GetCustomAttribute<LanguageServerEndpointAttribute>();
+            }
+
+            throw new InvalidOperationException("Somehow we are missing the method for our registered handler");
+
+            bool MethodInfoMatches(MethodInfo methodInfo)
+            {
+                var methodName = methodInfo.Name;
+
+                switch (requestType != null, responseType != null)
+                {
+                    case (true, true):
+                        return NameMatches(methodInfo, HandleRequestName, HandleRequestSuffix)
+                            && TypesMatch(methodInfo, [requestType, contextType, typeof(CancellationToken)]);
+
+                    case (false, true):
+                        return NameMatches(methodInfo, HandleRequestName, HandleNotificationSuffix)
+                            && TypesMatch(methodInfo, [contextType, typeof(CancellationToken)]);
+
+                    case (true, false):
+                        return NameMatches(methodInfo, HandleNotificationName, HandleNotificationSuffix)
+                            && TypesMatch(methodInfo, [requestType, contextType, typeof(CancellationToken)]);
+
+                    case (false, false):
+                        return NameMatches(methodInfo, HandleNotificationName, HandleNotificationSuffix)
+                            && TypesMatch(methodInfo, [contextType, typeof(CancellationToken)]);
+                }
+            }
+
+            static bool NameMatches(MethodInfo methodInfo, string name, string suffix)
+            {
+                var methodName = methodInfo.Name;
+
+                return methodName == name || methodName.EndsWith(suffix);
+            }
+
+            static bool TypesMatch(MethodInfo methodInfo, Type?[] types)
+            {
+                var parameters = methodInfo.GetParameters();
+                if (parameters.Length != types.Length)
+                    return false;
+
+                for (int i = 0, n = parameters.Length; i < n; i++)
+                {
+                    if (!Equals(types[i], parameters[i].ParameterType))
+                        return false;
+                }
+
+                return true;
+            }
+        }
+
+        static LanguageServerEndpointAttribute? GetMethodAttributeFromClassOrInterface(Type type)
+        {
+            var attribute = Attribute.GetCustomAttribute(type, typeof(LanguageServerEndpointAttribute)) as LanguageServerEndpointAttribute;
+
+            if (attribute is null)
+            {
+                var interfaces = type.GetInterfaces();
+                foreach (var @interface in interfaces)
+                {
+                    attribute = GetMethodAttributeFromClassOrInterface(@interface);
```

Comment:
interesting, is this something used on the Razor side?  I don't think in Roslyn we define method attributes on interfaces that a handler implements

Reply:
No idea! :smile: It came from here: https://github.com/dotnet/roslyn/blob/5f778bde9a17c34480dc83f3a17bf4cd9277d630/src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/HandlerProvider.cs#L188-L206

Reply:
Just found a place where we do this! https://github.com/dotnet/razor/blob/main/src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/EndpointContracts/IDidChangeConfigurationEndpoint.cs#L8

No reason I can think of we should keep it long term

Reply:
No idea I'm afraid. I guess being a source package, maybe it doesn't matter, and you can leave it out and we can just respond in time? I would imagine it would at least mean servers start seeing "method not found" errors, so they wouldn't be _completely_ in the dark.

Reply:
I kind of feel like this is outside of the scope of this PR. Probably should have an issue filed for later.

Reply:
yeah I didn't realize this was just moved.  fine with it as-is.